File: src/io/github/JakeJMattson/pixeldetails/ActionPanel.java
package io.github.JakeJMattson.pixeldetails;

import javax.swing.*;
import java.awt.*;

class ActionPanel extends JPanel
{
	private String label;
	private JLabel dynamicLabel;
	private PixelReader.Action action;

	ActionPanel(String label, PixelReader.Action action)
	{
		super(new FlowLayout(FlowLayout.LEFT));

		this.label = label;
		this.action = action;

		createPanel();
	}

	private void createPanel()
	{
		//Create labels
		JLabel staticLabel = new JLabel(label);
		staticLabel.setFont(new Font("Monospaced", Font.BOLD, 12));
		dynamicLabel = new JLabel();

		this.add(staticLabel);
		this.add(dynamicLabel);
	}

	String getLabel()
	{
		return label;
	}

	String getText()
	{
		return dynamicLabel.getText();
	}

	void performAction(Color pixelColor, Point mouse)
	{
		dynamicLabel.setText(action.performAction(pixelColor, mouse));
	}
}

File: src/main/kotlin/io/github/jakejmattson/pixeldetails/ActionPanel.kt
package io.github.jakejmattson.pixeldetails

import java.awt.*
import javax.swing.*

abstract class ActionPanel(val labelText: String) : JPanel(FlowLayout(FlowLayout.LEFT)) {
	private var dynamicLabel = JLabel()

	var text: String
		get() = dynamicLabel.text
		set(value) {dynamicLabel.text = value}

	init {
		add(JLabel(labelText).apply { font = Font("Monospaced", Font.BOLD, 12) })
		add(dynamicLabel)
	}

	abstract fun performAction(data: Any)
}

class CoordinatePanel(labelText: String, private val action: (mouse: Point) -> String) : ActionPanel(labelText) {
	override fun performAction(data: Any) {
		text = action.invoke(data as Point)
	}
}

class ColorPanel(labelText: String, private val action: (pixelColor: Color) -> String) : ActionPanel(labelText) {
	override fun performAction(data: Any) {
		text = action.invoke(data as Color)
	}
}